packages/flutter/lib/src/animation/curves.dart

```diff
@@ -267,6 +273,526 @@ class Cubic extends Curve {
+/// points.
+///
+/// {@template flutter.animation.curves.catmull_rom_description}
+/// Unlike most cubic splines, Catmull-Rom splines have the advantage that their
+/// curves pass through the control points given to them. They are cubic
+/// polynomial representations, and, in fact, Catmull-Rom splines can be
+/// converted mathematically into cubic splines. This class implements a
+/// "centripetal" Catmull-Rom spline implementation. The term centripetal
+/// implies that it won't form loops or self-intersections within a single
+/// segment.
+/// {@endtemplate}
+///
+/// See also:
+///
+///  * A Wikipedia article on [centripetal Catmull-Rom splines](https://en.wikipedia.org/wiki/Centripetal_Catmull%E2%80%93Rom_spline).
+///  * This [paper on using Catmull-Rom splines](http://faculty.cs.tamu.edu/schaefer/research/cr_cad.pdf).
+///  * [CatmullRomCurve], an animation curve that uses a [CatmullRomSpline] as its
+///    representation.
+class CatmullRomSpline extends Curve2D {
+  /// Constructs a centripetal Catmull-Rom spline curve.
+  ///
+  /// The `points` argument is a list of four or more points that describe the
+  /// points that the curve must pass through.
+  ///
+  /// The optional `tension` argument controls how tightly the curve approaches
+  /// the given `points`. It must be in the range 0.0 to 1.0, inclusive. It
+  /// defaults to 0.0, which provides the smoothest curve. A value of 1.0
+  /// produces a linear interpolation between points.
+  ///
+  /// The optional `endHandle` and `startHandle` points are the beginning and
+  /// ending handle positions. If not specified, they are created automatically
+  /// by extending the line formed by the first and/or last line segment in the
+  /// input points, respectively.
+  ///
+  /// The `tension` and `points` arguments must not be null, and the `points`
+  /// list must contain at least four points to interpolate.
+  CatmullRomSpline(
+      List<Offset> controlPoints, {
+        double tension = 0.0,
+        Offset startHandle,
+        Offset endHandle,
+      })  : assert(controlPoints != null),
+        assert(tension != null),
+        assert(tension <= 1.0, 'tension $tension must not be greater than 1.0.'),
+        assert(tension >= 0.0, 'tension $tension must not be negative.'),
+        assert(controlPoints.length > 3, 'There must be at least four control points to create a CatmullRomSpline.') {
+    // If not specified, select the first and last control points (which are
+    // handles: they are not intersected by the resulting curve) so that they
+    // extend the first and last segments, respectively.
+    startHandle ??= controlPoints[0] * 2.0 - controlPoints[1];
+    endHandle ??= controlPoints.last * 2.0 - controlPoints[controlPoints.length - 2];
+    final List<Offset> allPoints = <Offset>[
+      startHandle,
+      ...controlPoints,
+      endHandle,
+    ];
+
+    // An alpha of 0.5 is what makes it a centripetal Catmull-Rom spline. A
```

Comment:
Wondering whether we should do this computation lazily the first time we need it.